services/src/main/scala/cromwell/pipeline/service/AuthService.scala

```diff
@@ -26,16 +20,22 @@ class AuthService(userRepository: UserRepository, authUtils: AuthUtils)(implicit
 
   import authUtils._
 
-  def signIn(request: SignInRequest): Future[Option[AuthResponse]] =
-    OptionT(userRepository.getUserByEmail(request.email))
-      .filter(user => user.passwordHash == StringUtils.calculatePasswordHash(request.password.value, user.passwordSalt))
-      .map { user =>
-        val accessTokenContent = AccessTokenContent(user.userId)
-        val refreshTokenContent = RefreshTokenContent(user.userId, None)
-        getAuthResponse(accessTokenContent, refreshTokenContent, Instant.now.getEpochSecond)
+  def signIn(request: SignInRequest): Future[Option[AuthResponse]] = responseFromOptUser(authFilter(request))
+
+  def authFilter(request: SignInRequest): OptionT[Future, User] =
+    OptionT(userRepository.getUserByEmail(request.email)).semiflatMap[User] { user =>
+      passwordCorrect(request, user) match {
+        case true  => Future(user)
+        case false => Future.failed(IncorrectPasswordException("Wrong password"))
```

Comment:
In my opinion, if-else is more appropriate here. Base on [it](https://stackoverflow.com/questions/9266822/pattern-matching-vs-if-else)

Reply:
Nice advice! done! )